---
title: "SPLINTR barcode analysis, JFG34"
author: "Julia F Garcia, adapted from Dane Vass"
output: 
  rmarkdown::html_document:
    toc: true
    number_sections: true
    toc_depth: 3
    highlight: tango 
date: April 2024
editor_options: 
  markdown: 
    wrap: sentence
---

# Bartools Analysis of JFG34

Following the tutorial from: https://danevass.github.io/bartools/articles/bartools_quickstart.html
Note that I deleted some parts which I don't think are relevant for this experiment, but do check the tutorial in case they are important.

## 0. Install and load the `bartools` package

You can install bartools from [GitHub](https://github.com/DaneVass/bartools):
If you install using .renv, you will need to install again in every new RProject.

```{r eval=FALSE}
# first install Bioconductor dependencies
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(c("edgeR", "limma", "ComplexHeatmap"))

# then install bartools via GitHub
if (!requireNamespace("devtools", quietly = TRUE)) {
  install.packages("devtools")
}
devtools::install_github("DaneVass/bartools", dependencies = TRUE, force = TRUE)
```

```{r setup, warning=FALSE, message=FALSE}
# Don't inclue warning and messages in output
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
# Time stamp
Sys.time()

# Clean environment
rm(list=ls())
graphics.off()

#load libraries
library(bartools)
library(dplyr)
library(rmarkdown)
library(stringr)
knitr::opts_chunk$set(dev="png")

#results folder
results.plots <- "results/R/plots"

dir.create(results.plots, showWarnings = F, recursive = T)
```

## 1. Importing DNA barcode count data

### Generating a DGEList object from sample counts and metadata

Need to input following files:

-    txt files with barcode counts for each sample (output of BARTab pipeline)

-    a metadata .csv file with sample information, in this case I prepared the file with relevant information

```{r}
#Load metadata

samplesheet <- read.csv("data/R/metadata.csv",
  header = T,
  stringsAsFactors = F
)

samplesheet[samplesheet == ""] <- NA
samplesheet$files <- paste(samplesheet$X, "_rawcounts.txt", sep = "")

# Replace DOX/UT with +DOX/-DOX
columns_to_replace <- c("Name", "Treatment", "Biological_rep_2")

for (col in columns_to_replace) {
  samplesheet[[col]] <- gsub("^DOX", "+DOX", samplesheet[[col]])
  samplesheet[[col]] <- gsub("^UT", "-DOX",samplesheet[[col]])
}

paged_table(samplesheet)

```

Load in the counts as specified in the samplesheet into a DGEList object.
The function expects file locations to be specified either as character vector of filenames, or as a column named files in the samplesheet.
This results in the creation of a DGEList object containing counts and metadata information for each sample.
Check that all metadata information is present.

```{r}

test.dge <- edgeR::readDGE(
  files = samplesheet,
  path = "./data/R",
  group = samplesheet$Treatment,
  labels = samplesheet$Name,
  header = F
)

paged_table(test.dge$samples)

counts <- as.data.frame(test.dge$counts)
paged_table(counts)
```

### Separating samples

I will separate the 200k founder population as it has different barcodes to other samples, and having these present together could affect downstream analysis (?).

```{r}
#dataset with excluded 200k founder sample
samples_to_exclude <- c("f200k_A", "f200k_B")
test.dge <- test.dge[,!(test.dge$samples$Name %in% samples_to_exclude)]

```

## 2. Data QC and Normalisation

### Data QC - don't quite understand this

We first want to ensure that we are working with clean data.
Using the `thresholdCounts()` function we can determine an appropriate threshold to apply to the data to maximise signal to noise and retain true and informative barcodes.

Basically, deciding how to filter out barcodes that are present at low levels.
How many of each barcode needs to be present in X samples.

We can test different thresholding parameters, such as absolute thresholds on total read counts as below.

```{r}
# Remove barcodes (rows) with no data
test.dge <- test.dge[rowSums(test.dge$counts) != 0, ]
```

```{r}
thresholdCounts(
  test.dge,
  type = "absolute",
  threshold = 1,
  minSamples = 1,
  plot = T,
  group = "group"
)
thresholdCounts(
  test.dge,
  type = "absolute",
  threshold = 10,
  minSamples = 1,
  plot = T,
  group = "group"
)
thresholdCounts(
  test.dge,
  type = "absolute",
  threshold = 20,
  minSamples = 1,
  plot = T,
  group = "group"
)
thresholdCounts(
  test.dge,
  type = "absolute",
  threshold = 50,
  minSamples = 1,
  plot = T,
  group = "group"
)
thresholdCounts(
  test.dge,
  type = "absolute",
  threshold = 10,
  minSamples = 3,
  plot = T,
  group = "group"
)
```

Or relative thresholds based on proportion within a sample.

```{r}
thresholdCounts(
  test.dge,
  type = "relative",
  threshold = 1e-10,
  minSamples = 1,
  plot = T,
  group = "group"
)
thresholdCounts(
  test.dge,
  type = "relative",
  threshold = 1e-5,
  minSamples = 1,
  plot = T,
  group = "group"
)
thresholdCounts(
  test.dge,
  type = "relative",
  threshold = 1e-5,
  minSamples = 3,
  plot = T,
  group = "group"
)
```

Here we will continue with an absolute threshold of 20.

```{r}
dge.filtered <-
  thresholdCounts(
    test.dge,
    type = "absolute",
    threshold = 20,
    minSamples = 2,
    plot = F
  )
```

We then normalise samples to sequencing depth to counts per million using `normaliseCounts()`.

```{r}
dge.cpmnorm <- normaliseCounts(dge.filtered, method = "CPM")
```

We can plot the raw and normalised sequencing depth to get an idea of depth discrepancies between PCR replicates.

```{r}
# raw counts per sample
plotReadCounts(dge.filtered, group = "group")
```

```{r}
# normalised counts per sample
plotReadCounts(dge.cpmnorm, group = "group")
```

For lentiviral based cellular barcoding experiments, such as this one, it is common for the library to exhibit a degree of skewness based on the cloning method.
This means that some barcodes are represented in the library more than others and so have a greater chance to be transduced into multiple cells.\

Most experiments assume that each individual barcode is transduced into only one cell, and that each cell is only transduced with one barcode.
This is ensured using a low multiplicity of infection (MOI) transduction in which the likelihood that a cell is transduced with one or more barcode containing virions follows a Poisson distribution.

With this in mind, it also can be useful to check the total counts per barcode to identify bias in counts in sample vs. frequency of barcode in reference library.\
The barcodes are labelled based on their ranked frequency in the reference library, i.e. barcode 1 is the most frequent in the library and so on.

```{r}
# plot detected barcodes ordered by frequency in reference library
pdf(paste0(results.plots, "/", "frequency.pdf"), width = 8, height = 6)
plotBarcodeCounts(dge.cpmnorm, log10 = F)
dev.off()

# plot log10 barcode counts
plotBarcodeCounts(dge.cpmnorm, log10 = T)

# order barcodes by count across samples
plotBarcodeCounts(dge.cpmnorm, log10 = F, order = T)

# order barcodes by count across samples with log norm
plotBarcodeCounts(dge.cpmnorm, log10 = T, order = T)
```

In the first and second plot individual barcodes on the x-axis are ordered based on their frequency in the reference library pool.\
An increased number of counts per barcode toward the left hand side of the plot would be suggestive of transduction bias, meaning that there are more reads on average attributed to the more abundant barcodes in the library.
And so, likely multiple cells were transduced with the same barcode.\
We don't see this here suggesting that this is not a problem for this experiment.\
The third and fourth plot show the frequency distribution of barcodes in the experiment.


### Check correlation between PCR replicates

It is also important to ensure that individual samples are sequenced to an appropriate depth as this ensures that the entire barcode repertoire present in a sample is captured in the data.
Sequencing technical duplicates of a sample generated at the library PCR stage is a good way to ensure this.

```{r, fig.width=15, fig.height=12}
dge.filtered$samples$Name_replicates <- substr(dge.filtered$samples$Name, 1, nchar(test.dge$samples$Name) - 1)

# get all unique samples
# column "group" contains information on replicates here
unique_samples <- unique(dge.filtered$samples$Name_replicates)

library(gridExtra)  # Load gridExtra package for grid.arrange()

# Initialize an empty list to store plots
plot_list <- list()

# Iterate over each unique sample
for (x in unique_samples) {
  # Subset dge object to get replicates of current sample
  replicate_names <- colnames(dge.filtered)[dge.filtered$samples$Name_replicates %in% as.character(x)]
  
  # Check if there are at least two replicate names to plot
  if (length(replicate_names) >= 2) {
    # Generate plot using plotBarcodeRegression function
    plot <- plotBarcodeRegression(
      dge.filtered,
      sample1 = replicate_names[1],
      sample2 = replicate_names[2],
      rug = TRUE,
      trans = "log1p"
    )
    
    # Add the plot to the plot_list
    plot_list[[length(plot_list) + 1]] <- plot
  } else {
    # Print a message if less than two replicates are found for the sample
    cat(paste("Less than two replicates found for sample:", x, "\n"))
  }
}

if (length(plot_list) > 0) {
  grid.arrange(grobs = plot_list, ncol = 3)  # Adjust 'ncol' based on desired layout
} else {
  cat("No plots to arrange.\n")
}

# Arrange plots in a grid layout using grid.arrange
pdf(paste0(results.plots, "/", "pcr_replicates.pdf"), width = 15, height = 12)


if (length(plot_list) > 0) {
  grid.arrange(grobs = plot_list, ncol = 3)  # Adjust 'ncol' based on desired layout
} else {
  cat("No plots to arrange.\n")
}

dev.off()


```

We fit a linear model to both technical replicates per sample and plot the regression line.
Note that we expect a very high correlation because these are PCR replicates of the same barcode pool.

We can also easily get the correlation values between replicates using `calcReplicateCorr`.

Samples can be filtered for high or low correlation using the `threshold` and `return` variables.

```{r}
corrs <- calcReplicateCorr(dge.filtered, group = "Name_replicates")
#corrs[which(corrs < 0.999)]
corrs
```


### Collapse PCR replicates in object

Now that we know our samples are of good quality we have no further use of the PCR replicate information.
From this point onward its a good idea to collapse our PCR replicates.

```{r}
dim(dge.filtered)
```

`collapseReplicates` can take the average (default behavior) or the sum of PCR technical replicates within each sample.
Here we take the average.
Users may want to sum PCR replicates if there is evidence of sampling bias across technical repeats (i.e. poor correlation score or other evidence).

```{r}
dge.filtered.collapsed <- collapseReplicates(
  dge.filtered,
  group = "Name_replicates",
  method = "mean"
)
```

The result is a clean barcode sequencing dataset ready for further investigation and visualisation.

```{r eval=FALSE}
head(dge.filtered.collapsed)
```

## 3. Visualisation

`bartools` includes a range of visualisation options for examining barcode-seq datasets.

### Prepare data and further exclude samples

I may need to exclude samples that are not good enough, i.e. library prep was not good; only one PCR replicate; sequencing QC not looking good. I made an excel file with this information to decide which samples to exclude - in future could add this to the metadata and exclude in R automatically. 

```{r}
#Prepare sheet

# Order samples

new_order <- c("f100k_", "-DOX_day05_1", "-DOX_day05_2", "-DOX_day05_3", "-DOX_day09_1", "-DOX_day09_2", "-DOX_day09_3", "-DOX_day13_1", "-DOX_day13_3", "+DOX_day05_1", "+DOX_day05_2", "+DOX_day09_1", "+DOX_day09_2","+DOX_day13_1", "+DOX_day13_2", "+DOX_day13_3") 

dge.filtered.collapsed$counts <- dge.filtered.collapsed$counts[, new_order]

dge.filtered.collapsed$samples <- dge.filtered.collapsed$samples[new_order, ]

# Exclude samples 

# Identify column names to exclude
samples_to_exclude <- c("f100k_", "-DOX_day05_1", "-DOX_day05_2", "-DOX_day09_1", "-DOX_day13_1","+DOX_day13_2", "+DOX_13_3")

dge.filtered.collapsed.excluded <- dge.filtered.collapsed

dge.filtered.collapsed.excluded$counts <- dge.filtered.collapsed$counts[, !colnames(dge.filtered.collapsed$counts) %in% samples_to_exclude]

dge.filtered.collapsed.excluded$samples <- dge.filtered.collapsed$samples[!rownames(dge.filtered.collapsed$samples) %in% samples_to_exclude,]

#Further separate samples into DOX & UT that have all 3 time-points

samples_to_include <- c("+DOX_day05_1", "+DOX_day09_1", "+DOX_day13_1")

dge.filtered.collapsed.dox <- dge.filtered.collapsed
dge.filtered.collapsed.dox$counts <- dge.filtered.collapsed$counts[, colnames(dge.filtered.collapsed$counts) %in% samples_to_include]
dge.filtered.collapsed.dox$samples <- dge.filtered.collapsed$samples[rownames(dge.filtered.collapsed$samples) %in% samples_to_include,]

samples_to_include <- c("-DOX_day05_3", "-DOX_day09_3", "-DOX_day13_3")

dge.filtered.collapsed.ut <- dge.filtered.collapsed
dge.filtered.collapsed.ut$counts <- dge.filtered.collapsed$counts[, colnames(dge.filtered.collapsed$counts) %in% samples_to_include]
dge.filtered.collapsed.ut$samples <- dge.filtered.collapsed$samples[rownames(dge.filtered.collapsed$samples) %in% samples_to_include,]


#Day 13 -DOx vs +DOZ
samples_to_include <- c("+DOX_day13_1", "-DOX_day13_3")

dge.filtered.collapsed.13 <- dge.filtered.collapsed
dge.filtered.collapsed.13$counts <- dge.filtered.collapsed$counts[, colnames(dge.filtered.collapsed$counts) %in% samples_to_include]
dge.filtered.collapsed.13$samples <- dge.filtered.collapsed$samples[rownames(dge.filtered.collapsed$samples) %in% samples_to_include,]

```


### Number of barcodes

```{r, fig.width=8, fig.height=6}

plotDetectedBarcodes(
  dge.filtered.collapsed,
  percentile = 1,
  plot = T,
  group = "Treatment", 
  title = "Number of barcodes detected - All samples"
) + 
  theme(axis.text.x=element_text(size=10))



plot <- plotDetectedBarcodes(
  dge.filtered.collapsed.excluded,
  percentile = 1,
  plot = T,
  group = "Treatment", 
  title = "Number of barcodes detected"
) + 
  theme(axis.text.x=element_text(size=10))
  
plot

pdf(paste0(results.plots, "/", "nr_of_barcodes.pdf"), width = 8, height = 7)
print(plot)
dev.off()
```


### Principal Components Analysis

A global level PCA analysis is a good way to get a high level understanding of the similarities and differences between samples.

```{r, fig.width= 15, fig.height=4.5}
library(gridExtra)  # Load gridExtra package for grid.arrange()

# Initialize an empty list to store plots
plot_list <- list()

# Plot Barcode PCA for Treatment groups
plot1 <- plotBarcodePCA(dge.filtered.collapsed.excluded, 
                        groups = "Treatment", 
                        ntop = 1000,
                        pcs = c(1, 2))
plot_list[[length(plot_list) + 1]] <- plot1

# Plot Barcode PCA for Timepoint groups
plot2 <- plotBarcodePCA(dge.filtered.collapsed.excluded, 
                        groups = "Timepoint", 
                        ntop = 1000,
                        pcs = c(1, 2))
plot_list[[length(plot_list) + 1]] <- plot2

# Plot Barcode PCA for Timepoint_2 groups
plot3 <- plotBarcodePCA(dge.filtered.collapsed.excluded, 
                         groups = "Timepoint_2", 
                         ntop = 1000,
                         pcs = c(1, 2))
plot_list[[length(plot_list) + 1]] <- plot3

grid.arrange(grobs = plot_list, ncol = 3)


# Arrange plots in a grid layout using grid.arrange
pdf(paste0(results.plots, "/", "pca.pdf"), width = 15, height = 4.5)
grid.arrange(grobs = plot_list, ncol = 3)
dev.off()

```

It looks like there is some clustering based on -/+ DOX, and also based on time-point. But biological replicates don't cluster so well, so maybe this clustering is not so informative?

### Bubble plot

Here, barcodes/tags are represented by bubbles aligned on a single plane.
The size of the bubbles reflects the percentage abundance of each barcode within a sample.

First, we look at all samples together, including the non-excluded ones. Can see that in some samples there are some barcodes that are at a much higher proportion, which in this experiment may not represent the truth as they differ from their biological replicates. In the founder population (f100k) we should expect quite an even distribution (as seen in f200k), and this is not seen here, suggesting the sequencing wasn't very accurate (as we know is true).


```{r}
plotBarcodeBubble(dge.filtered.collapsed, 
                  proportionCutoff = 10, 
                  labelBarcodes = T, 
                  group = "Treatment", 
                  title = "All samples")
```


Now if we exclude samples that we do not trust the sequencing, we see a much more even distribution of barcode sequencing. A few conclusions can be drawn here: 
 - Barcode frequency distribution looks quite similar across samples, suggesting that there isn't much difference in -/+ DOX, but also that the evolution of these cells always is quite similar suggesting deterministic evolution.  
 - There doesn't seem to be dominant clones in any samples or time-points, but it looks like at later time-points some clones become more frequent than others.

```{r}
plot <- plotBarcodeBubble(dge.filtered.collapsed.excluded, 
                  proportionCutoff = 10, 
                  labelBarcodes = T, 
                  group = "Treatment")

plot
pdf(paste0(results.plots, "/", "bubble_plot.pdf"), width = 6, height = 6)
print(plot)
dev.off()



plotBarcodeBubble(dge.filtered.collapsed.excluded, 
                  proportionCutoff = 10, 
                  labelBarcodes = T, 
                  group = "Timepoint", 
                  title = "Grouped by time-point")


```


Only 1 biological rep for -DOX and one for +DOX have all time-points included. If we look at these two, we can see that indeed it seems some barcodes become more prominent at day 13. 

Using the `orderSample` parameter, bubbleplots can also be arranged according to frequency in a particular sample which can help with visual comparison of large vs small clones across samples and conditions.

```{r}

plotBarcodeBubble(dge.filtered.collapsed.ut, 
                  proportionCutoff = 10, 
                  labelBarcodes = T, 
                  group = "Treatment", 
                  title = "-DOX")



plotBarcodeBubble(dge.filtered.collapsed.dox, 
                  proportionCutoff = 10, 
                  labelBarcodes = T, 
                  group = "Treatment", 
                  title = "+DOX")


```
This can also be seen in a timeseries plot:

### Timeseries Plots

We can use `plotBarcodeTimeseries` to get an idea of the relative abundance of the top *n* barcodes in a sample relative to others. In this case I am looking at the top 20 most abundant clones - of all samples?

```{r}

plot <- plotBarcodeTimeseries(dge.filtered.collapsed.13, 
                      top = 20, 
                      seed = 10101) +
  ggtitle("-DOX vs +DOX day 13")

plot
pdf(paste0(results.plots, "/", "time_series_dox_ut_13.pdf"), width = 6, height = 6)
print(plot)
dev.off()

plot <- plotBarcodeTimeseries(dge.filtered.collapsed.ut, 
                      top = 20, 
                      seed = 10101) +
  ggtitle("-DOX over time") +
  ylim(0, 50)

plot
pdf(paste0(results.plots, "/", "time_series_ut.pdf"), width = 6, height = 6)
print(plot)
dev.off()

plot <- plotBarcodeTimeseries(dge.filtered.collapsed.dox, 
                      top = 20, 
                      seed = 10101) +
  ggtitle("+DOX over time") +
  ylim(0, 50)

plot
pdf(paste0(results.plots, "/", "time_series_dox.pdf"), width = 6, height = 6)
print(plot)
dev.off()
```


### Barcode Plot

Alternatively, we can focus in on the most abundant barcodes within a set of samples to more easily observe how these change in frequency over the course of an experiment.

```{r}
plotBarcodeHistogram(dge.filtered.collapsed, topN = 20, alphaLowFreq = 0)

plotBarcodeHistogram(dge.filtered.collapsed.excluded, topN = 20, alphaLowFreq = 0)

plotBarcodeHistogram(dge.filtered.collapsed.ut, topN = 20, alphaLowFreq = 0)

plotBarcodeHistogram(dge.filtered.collapsed.dox, topN = 20, alphaLowFreq = 0)

plotBarcodeHistogram(dge.filtered.collapsed.13, topN = 20, alphaLowFreq = 0)



```

```{r eval = FALSE}
#Check if replicates are similar (probably not the best way to check)
plotBarcodeHistogram(dge.filtered.collapsed.excluded, topN = 20, alphaLowFreq = 0, samples = c("UT_day09_2", "UT_day09_3"))
plotBarcodeHistogram(dge.filtered.collapsed.excluded, topN = 20, alphaLowFreq = 0, samples = c("DOX_day05_1", "DOX_day05_2"))
plotBarcodeHistogram(dge.filtered.collapsed.excluded, topN = 20, alphaLowFreq = 0, samples = c("DOX_day09_1", "DOX_day09_2"))
plotBarcodeHistogram(dge.filtered.collapsed.excluded, topN = 20, alphaLowFreq = 0, samples = c("DOX_day13_1", "DOX_day13_3"))
```


### Heatmaps

Another method of comparing abundance across samples is using a heatmap.
Here barcodes ranked among the top n most abundant within each sample are indicated by an asterisk.

```{r}
plotBarcodeHeatmap(
  normaliseCounts(dge.filtered.collapsed.excluded, method = "CPM"), 
  topN = 15,
  showBarcodes = F,
  group = "Timepoint"
)


plotBarcodeHeatmap(
  normaliseCounts(dge.filtered.collapsed.ut, method = "CPM"), 
  topN = 15,
  showBarcodes = F,
  group = "Timepoint"
)

plotBarcodeHeatmap(
  normaliseCounts(dge.filtered.collapsed.dox, method = "CPM"), 
  topN = 15,
  showBarcodes = F,
  group = "Timepoint"
)

plotBarcodeHeatmap(
  normaliseCounts(dge.filtered.collapsed.13, method = "CPM"), 
  topN = 15,
  showBarcodes = F
)
```

## 4. Analysing Composition and Diversity

Its important to not only be able to visualise the data but also understand relationships between barcodes/tags at the data level.

### Identifying abundant barcodes within samples

It is important to be able to determine which barcodes are most abundant within each sample.
`bartools` allows this to be easily calculated according to an abundance threshold.

And then we can plot venn diagrams to see how similar the replicates are.

```{r, fig.width=15, fig.height=6}
#Get top barcodes in each sample, can play around with the threshold (right now at 0.1%)
top.bc <- getDominantBarcodes(dge.filtered.collapsed.excluded, threshold = 0.001)

library(VennDiagram)
library(gridExtra)

venn1 <- venn.diagram(
  x = list(rep2 = top.bc$`-DOX_day09_2`, rep3 = top.bc$`-DOX_day09_3`),
  filename = NULL,
  category.names = c("rep2", "rep3"),
  col=c("#cc99cc", '#21908dff'),
  fill = c(alpha("#cc99cc",0.3), alpha('#21908dff',0.3)),
  cex = 2,
  fontfamily = "sans", 
  cat.fontfamily = "sans", 
  cat.pos = c(-35, 27),
  cat.dist = c(0.035, 0.035),
  main = "UT day 9",
  main.fontfamily = "sans", 
  main.cex = 2,
  cat.cex = 2, 
  lwd = 5

)   

venn2 <- venn.diagram(
  x = list(rep2 = top.bc$`+DOX_day05_1`, rep3 = top.bc$`+DOX_day05_1`),
  filename = NULL,
  category.names = c("rep1", "rep2"),
  col=c("#cc99cc", '#21908dff'),
  fill = c(alpha("#cc99cc",0.3), alpha('#21908dff',0.3)),
  cex = 2,
  fontfamily = "sans", 
  cat.fontfamily = "sans", 
  cat.pos = c(-35, 27),
  cat.dist = c(0.035, 0.035),
  main = "DOX day 5",
  main.fontfamily = "sans", 
  main.cex = 2,
  cat.cex = 2, 
  lwd = 5

) 

venn3 <- venn.diagram(
  x = list(rep2 = top.bc$`+DOX_day09_1`, rep3 = top.bc$`+DOX_day09_2`),
  filename = NULL,
  category.names = c("rep1", "rep2"),
  col=c("#cc99cc", '#21908dff'),
  fill = c(alpha("#cc99cc",0.3), alpha('#21908dff',0.3)),
  cex = 2,
  fontfamily = "sans", 
  cat.fontfamily = "sans", 
  cat.pos = c(-35, 27),
  cat.dist = c(0.035, 0.035),
  main = "DOX day 9",
  main.fontfamily = "sans", 
  main.cex = 2,
  cat.cex = 2, 
  lwd = 5

) 

venn4 <- venn.diagram(
  x = list(rep2 = top.bc$`+DOX_day13_1`, rep3 = top.bc$`+DOX_day13_3`),
  filename = NULL,
  category.names = c("rep1", "rep3"),
  col=c("#cc99cc", '#21908dff'),
  fill = c(alpha("#cc99cc",0.3), alpha('#21908dff',0.3)),
  cex = 2,
  fontfamily = "sans", 
  cat.fontfamily = "sans", 
  cat.pos = c(-35, 27),
  cat.dist = c(0.035, 0.035),
  main = "DOX day 13",
  main.fontfamily = "sans", 
  main.cex = 2,
  cat.cex = 2, 
  lwd = 5

) 

#grid.arrange(venn1)

grid.arrange(venn2, venn3, venn4, ncol = 3)

pdf(paste0(results.plots, "/", "biological_rep_overlap.pdf"), width = 15, height = 6)
grid.arrange(venn2, venn3, venn4, ncol = 3)
dev.off()
```


We can then use specific plots to visualise the dominance of specific barcodes within and across samples.

```{r, fig.width=10, fig.height=8}
# plot top barcodes across samples 
# in the replicates
top.bc <- getDominantBarcodes(dge.filtered.collapsed.excluded, threshold = 0.008)
top.bc

plotBarcodeBoxplot(
  dge.filtered.collapsed.excluded,
  barcodes = top.bc$`-DOX_day13_3`,
  group = "Biological_rep_2",
  conditions = c("-DOX_day13_rep3", "-DOX_day5_rep3", "-DOX_day9_rep3"),
  point = T
)


plotBarcodeBoxplot(
  dge.filtered.collapsed.excluded,
  barcodes = top.bc$`-DOX_day13_3`,
  group = "Biological_rep_2",
  conditions = c("-DOX_day13_rep3", "+DOX_day13_rep1", "+DOX_day13_rep3"),
  point = T
)


plotBarcodeBoxplot(
  dge.filtered.collapsed.excluded,
  barcodes = top.bc$`+DOX_day13_1`,
  group = "Biological_rep_2",
  conditions = c("-DOX_day13_rep3", "+DOX_day13_rep1", "+DOX_day13_rep3"),
  point = T
)
```

### Calculating and plotting percentile abundance.

The above graphs demonstrate that relatively few barcodes can sometimes comprise the majority of a sample's clonality, particularly following a selective event such as drug treatment.
It is useful to formally analyse this based on a desired percentile threshold.
A common threshold is the 95th percentile.
This can eliminate small barcodes that comprise the tail of the dataset and give a sense of how many clones truly comprise each sample

I think this is only relevant if there are dominant clones?

```{r, eval=FALSE}
top_barcodes <- calcPercentileBarcodes(dge.filtered.collapsed.excluded, percentile = 0.95)
top_barcodes$NumBarcodes
top_barcodes$TopBarcodeCounts$`UT_13_3`
top_barcodes$TopBarcodes$`UT_13_3`
```

We can compare the number of detected barcodes in the top 95th percentile per sample and the total sample.

```{r, eval=FALSE}
plotDetectedBarcodes(
  dge.filtered.collapsed.excluded,
  percentile = 1,
  plot = T,
  group = "Treatment", 
)
plotDetectedBarcodes(
  dge.filtered.collapsed.excluded,
  percentile = 0.95,
  plot = T,
  group = "Treatment"
)

plotDetectedBarcodes(
  dge.filtered.collapsed.excluded,
  percentile = 0.80,
  plot = T,
  group = "Treatment"
)
```

These plots show that there are few clones that comprise the majority of the dataset per mouse.
Also, there are generally fewer clones present in the high dose group compared to the vehicle or low dose groups.

### Diversity analysis

We can examine within-sample diversity in a few different ways.
The most common are Shannon, Simpson, Inverse Simpson and Gini.
Each will be applicable in different circumstances, however the Shannon diversity index is more widely used to compare global diversity amongst populations of barcoded cells.

`calcDivIndexes` can be used to determine various diversity indices per sample. These diversity calculations can then be fed to `plotDivIndexes` for visualisation as either a bar or dotplot

```{r}
diversity <- calcDivIndexes(dge.filtered.collapsed.excluded)
paged_table(diversity)

# bar plot
plotDivIndexes(
  dge.filtered.collapsed.excluded,
  div = diversity,
  metric = "shannon",
  group = "Treatment"
)


```




```{r}
diversity <- calcDivIndexes(dge.filtered.collapsed.ut)
paged_table(diversity)

plot <- plotDivIndexes(
  dge.filtered.collapsed.ut,
  div = diversity,
  metric = "shannon"
) 

plot
  
pdf(paste0(results.plots, "/", "diversity_ut.pdf"), width = 6, height = 6)
print(plot)
dev.off()

##########
diversity <- calcDivIndexes(dge.filtered.collapsed.dox)
paged_table(diversity)


plot <- plotDivIndexes(
  dge.filtered.collapsed.dox,
  div = diversity,
  metric = "shannon"
)

plot
pdf(paste0(results.plots, "/", "diversity_dox.pdf"), width = 6, height = 6)
print(plot)
dev.off()

#########
diversity <- calcDivIndexes(dge.filtered.collapsed.13)
paged_table(diversity)


plot <- plotDivIndexes(
  dge.filtered.collapsed.13,
  div = diversity,
  metric = "shannon"
)
plot

pdf(paste0(results.plots, "/", "diversity_utvsdox_13.pdf"), width = 6, height = 6)
print(plot)
dev.off()

```


### Comparing abundance

We can also statistically test for barcodes / tags that are over / underrepresented in a group of samples relative to another using the internal edgeR framework.
`bartools` contains a convenience wrapper for this functionality

```{r, fig.width=8, fig.height=8, eval=FALSE}

##not working need to change names because of the -/+ DOX
compareAbundance(dge.filtered.collapsed.excluded,
                 group = "Timepoint_2", 
                 condition1 = "UT_day13",
                 condition2 = "DOX_day13",
                 pval = 0.05,
                 logFC = 10)
```

## 5. Session Info

```{r}
sessionInfo()
```
